async/async.go

```diff
@@ -0,0 +1,372 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azeventgrid/publisher"
+	"github.com/RMI/pacta/blob"
+	"github.com/RMI/pacta/pacta"
+	"github.com/RMI/pacta/task"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+)
+
+type Config struct {
+	Blob   Blob
+	PubSub *publisher.Client
+	Logger *zap.Logger
+}
+
+func (c *Config) validate() error {
+	if c.Blob == nil {
+		return errors.New("no blob client given")
+	}
+	if c.PubSub == nil {
+		return errors.New("no pub/sub client given")
+	}
+	if c.Logger == nil {
+		return errors.New("no logger given")
+	}
+
+	return nil
+}
+
+type Blob interface {
+	ReadBlob(ctx context.Context, uri string) (io.ReadCloser, error)
+	WriteBlob(ctx context.Context, uri string, r io.Reader) error
+	Scheme() blob.Scheme
+}
+
+type Handler struct {
+	blob   Blob
+	pubsub *publisher.Client
+	logger *zap.Logger
+}
+
+func New(cfg *Config) (*Handler, error) {
+	if err := cfg.validate(); err != nil {
+		return nil, fmt.Errorf("invalid config given: %w", err)
+	}
+
+	return &Handler{
+		blob:   cfg.Blob,
+		pubsub: cfg.PubSub,
+		logger: cfg.Logger,
+	}, nil
+}
+
+// TODO: Send a notification when parsing fails.
+func (h *Handler) ParsePortfolio(ctx context.Context, taskID task.ID, req *task.ParsePortfolioRequest, destPortfolioContainer string) error {
+	// Load the portfolio from blob storage, place it in /mnt/raw_portfolios, where
+	// the `process_portfolios.R` script expects it to be.
+	for _, srcURI := range req.BlobURIs {
+		id := uuid.New().String()
+		// TODO: Probably set the CSV extension in the signed upload URL instead.
+		destPath := filepath.Join("/", "mnt", "raw_portfolios", fmt.Sprintf("%s.csv", id))
+		if err := h.downloadBlob(ctx, string(srcURI), destPath); err != nil {
+			return fmt.Errorf("failed to download raw portfolio blob: %w", err)
+		}
+	}
+
+	processedDir := filepath.Join("/", "mnt", "processed_portfolios")
+	if err := os.MkdirAll(processedDir, 0600); err != nil {
+		return fmt.Errorf("failed to create directory to download blob to: %w", err)
+	}
+
+	var stdout, stderr bytes.Buffer
+	cmd := exec.CommandContext(ctx, "/usr/local/bin/Rscript", "/app/process_portfolios.R")
+	cmd.Stdout = io.MultiWriter(os.Stdout, &stdout)
+	cmd.Stderr = io.MultiWriter(os.Stderr, &stderr)
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to run process_portfolios script: %w", err)
+	}
+
+	sc := bufio.NewScanner(&stderr)
+
+	// TODO: Load from the output database file (or similar, like reading the processed_portfolios dir) instead of parsing stderr
+	var paths []string
+	for sc.Scan() {
+		line := sc.Text()
+		idx := strings.Index(line, "writing to file: " /* 17 chars */)
+		if idx == -1 {
+			continue
+		}
+		paths = append(paths, strings.TrimSpace(line[idx+17:]))
+	}
+
+	// NOTE: This code could benefit from some concurrency, but I'm opting not to prematurely optimize.
+	var out []*task.ParsePortfolioResponseItem
+	for _, p := range paths {
+		lineCount, err := countCSVLines(p)
```

Comment:
Let's remove this - in the discussion with RMI on the doc I sent over last week we agreed that any metrics displayed to the user should come from them.

Reply:
Excellent news for you (which you've likely seen at this point): The updated parser code does surface this info, and this is indeed removed there.